kong/tools/queue.lua

```diff
@@ -0,0 +1,369 @@
+end
+
+local Queue = {}
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+---------
+-- Queue
+---------
+
+
+local queues = {}
+
+
+function Queue.exists(name)
+  return queues[name] and true or false
+end
+
+-------------------------------------------------------------------------------
+-- Initialize a queue with background retryable processing
+-- @param process function, invoked to process every payload generated
+-- @param opts table, requires `name`, optionally includes `retry_count`, `max_delay` and `batch_max_size`
+-- @return table: a Queue object.
+function Queue.get(name, handler, opts)
+
+  assert(type(name) == "string",
+    "arg #1 (name) must be a string")
+  assert(type(handler) == "function",
+    "arg #2 (handler) must be a function")
+  assert(type(opts) == "table",
+    "arg #3 (opts) must be a table")
+
+  assert(not opts.name or (opts.name == name),
+    "inconsistent queue name in argument and queue configuration")
+
+  local queue = queues[name]
+  if queue then
+    queue:log(DEBUG, "queue exists")
+    return queue
+  end
+
+  queue = {
+    name = name,
+    handler = handler,
+
+    semaphore = semaphore.new(),
+    retry_delay = 1,
+
+    running = true,
+    last_used = now(),
+    bytes_queued = 0,
+    queue = {},
+  }
+  for option, value in pairs(opts) do
+    queue[option] = value
+  end
+
+  queue = setmetatable(queue, Queue_mt)
+
+  kong.timer:named_every(name, queue.poll_time, function(premature, q)
+    if q.running
+      and not ngx.worker.exiting()
+      and (now() - q.last_used) < q.max_idle_time
+    then
+      q:log(DEBUG, "processing queue")
+      q:process_once(0)
+    else
+      if premature then
+        q:log(DEBUG, "shutting down queue")
+        q:drain()
+      else
+        kong.timer:cancel(name)
+      end
+      queues[name] = nil
+    end
+  end, queue)
+
+  queues[name] = queue
+
+  queue:log(DEBUG, "queue created")
+
+  return queue
+end
+
+
+-------------------------------------------------------------------------------
+-- Log a message that includes the name of the queue for identification purposes
+-- @param self Queue
+-- @param level: log level
+-- @param formatstring: format string, will get the queue name and ": " prepended
+-- @param ...: formatter arguments
+function Queue:log(level, formatstring, ...)
+  local message = self.name .. ": " .. formatstring
+  if select('#', ...) > 0 then
+    return ngx.log(level, string.format(message, unpack({...})))
+  else
+    return ngx.log(level, message)
+  end
+end
+
+
+function Queue:delete_frontmost_entry()
+  if self.string_capacity then
+    self.bytes_queued = self.bytes_queued - #self.queue[1]
+  end
+  table.remove(self.queue, 1)
+end
+
+
+function Queue:process_once(timeout)
+  local ok, err = self.semaphore:wait(timeout)
+  if not ok then
+    if err ~= "timeout" then
+      self:log(ERR, 'error waiting for semaphore: %s', err)
+    end
+    return
+  end
+  self.last_used = now()
+  local data_started = now()
```

Comment:
We should not call `now()` twice.

Reply:
It does not matter and is easier to read this way.